File: programmers/python/level2/kakao/2019/tuple.py
#https://programmers.co.kr/learn/courses/30/lessons/64065
def solution(s):
    s = s[2:-2].split("},{")
    tuple = []
    for c in s :
        tuple.append(list(map(int, c.split(","))))

    tuple.sort(key = lambda x: len(x))
    answer = []
    for t in tuple:
        if len(t) == 1:
            answer.append(int(t[0]))
        else:
            for x in t :
                x  = int(x)
                if x not in answer:
                    answer.append(x)
    return answer

from collections import deque
def solution_old(s):
    s = deque(s[1:-1])

    input_list = []
    while s:
        x = s.popleft()
        if x=='{' :
            str_tem = ''
            while True:
                y = s.popleft()
                if y == '}':
                    break
                str_tem += y
            input_list.append(set(map(int, str_tem.split(","))))
    input_list.sort(key=lambda x: len(x))
    answer = []
    now = set()
    for s in input_list:
        left = s - now
        answer.append(*left)
        now = s
    return answer

print(solution("{{2},{2,1},{2,1,3},{2,1,3,4}}"))
print(solution("{{1,2,3},{2,1},{1,2,4,3},{2}}"))
print(solution("{{20,111},{111}}"))
print(solution("{{123}}"))
print(solution("{{4,2,3},{3},{2,3,4,1},{2,3}}"))


def other(s):
    answer = []
    s1 = s.lstrip('{').rstrip('}').split('},{')
    new_s = []
    for i in s1:
        new_s.append(i.split(','))
    new_s.sort(key = len)
    for i in new_s:
        for j in range(len(i)):
            if int(i[j]) not in answer:
                answer.append(int(i[j]))
    return answer

print(other("{{2},{2,1},{2,1,3},{2,1,3,4}}"))

File: programmers/java/src/level2/kakao/kakao2019/Tuple.java
package level2.kakao.kakao2019;

import java.util.*;
import java.util.stream.Collectors;

// https://programmers.co.kr/learn/courses/30/lessons/64065
public class Tuple {
    public static void main(String[] args) {
        Tuple t = new Tuple();
        System.out.println(t.solution("{{2},{2,1},{2,1,3},{2,1,3,4}}"));
        System.out.println(t.solution("{{1,2,3},{2,1},{1,2,4,3},{2}}"));
        System.out.println(t.solution("{{20,111},{111}}"));
        System.out.println(t.solution("{{123}}"));
        System.out.println(t.solution("{{4,2,3},{3},{2,3,4,1},{2,3}}"));
        System.out.println(t.other("{{4,2,3},{3},{2,3,4,1},{2,3}}"));
    }

    private List<Integer> solution(String s) {
        s = s.substring(2, s.length()-2);
        String[] split = s.split("},\\{");
        List<List<Integer>> tuple = new ArrayList<>();
        for (String s1 : split) {
            List<Integer> numbers = new ArrayList<>();
            String[] split1 = s1.split(",");
            for (String s2 : split1) {
                numbers.add(Integer.parseInt(s2));
            }
            tuple.add(numbers);
        }
        tuple.sort(Comparator.comparingInt(List::size));
        List<Integer> answer = new ArrayList<>();
        for (List<Integer> t : tuple) {
            for (Integer integer : t) {
                if(answer.contains(integer)) {
                    continue;
                }
                answer.add(integer);
            }
        }
        return answer;
    }

    public List<Integer> solution_old(String s) {
        s = s.substring(1, s.length() - 2);
        List<Set<Integer>> list = new ArrayList<>();
        Arrays.stream(s.split("},"))
                .map(x -> x.substring(1))
                .map(x -> Arrays.stream(x.split(",")).map(Integer::parseInt).collect(Collectors.toSet()))
                .forEach(list::add);
        list.sort(Comparator.comparingInt(Set::size));
        Set<Integer> now = new HashSet<>();
        List<Integer> answer = new ArrayList<>();
        for (Set<Integer> integers : list) {
            integers.removeAll(now);
            Integer integer = integers.stream().findFirst().get();
            answer.add(integer);
            now.add(integer);
        }
        return answer;
    }

    public int[] other(String s) {
        Set<String> set = new HashSet<>();
        String[] arr = s.replaceAll("[{]", " ").replaceAll("[}]", " ").trim().split(" , ");
        Arrays.sort(arr, (a, b) -> {
            return a.length() - b.length();
        });
        int[] answer = new int[arr.length];
        int idx = 0;
        for (String s1 : arr) {
            for (String s2 : s1.split(",")) {
                if (set.add(s2)) answer[idx++] = Integer.parseInt(s2);
            }
        }
        return answer;
    }
}
